sdk/Pulumi/Deployment/Callbacks.cs

```diff
@@ -0,0 +1,173 @@
+
+        public CallbacksHost()
+        {
+            this._host = Host.CreateDefaultBuilder()
+                .ConfigureWebHostDefaults(webBuilder =>
+                {
+                    webBuilder
+                        .ConfigureKestrel(kestrelOptions =>
+                        {
+                            kestrelOptions.Listen(IPAddress.Loopback, 0, listenOptions =>
+                            {
+                                listenOptions.Protocols = HttpProtocols.Http2;
+                            });
+                        })
+                        .ConfigureAppConfiguration((context, config) =>
+                        {
+                            // clear so we don't read appsettings.json
+                            // note that we also won't read environment variables for config
+                            config.Sources.Clear();
+                        })
+                        .ConfigureLogging(loggingBuilder =>
+                        {
+                            // disable default logging
+                            loggingBuilder.ClearProviders();
+                        })
+                        .ConfigureServices(services =>
+                        {
+                            // Injected into Callbacks constructor
+                            services.AddSingleton(new Callbacks(_targetTcs.Task));
+
+                            services.AddGrpc(grpcOptions =>
+                            {
+                                // MaxRpcMesageSize raises the gRPC Max Message size from `4194304` (4mb) to `419430400` (400mb)
+                                var maxRpcMesageSize = 1024 * 1024 * 400;
+
+                                grpcOptions.MaxReceiveMessageSize = maxRpcMesageSize;
+                                grpcOptions.MaxSendMessageSize = maxRpcMesageSize;
+                            });
+                        })
+                        .Configure(app =>
+                        {
+                            app.UseRouting();
+                            app.UseEndpoints(endpoints =>
+                            {
+                                endpoints.MapGrpcService<Callbacks>();
+                            });
+                        });
+                })
+                .Build();
+
+            // before starting the host, set up this callback to tell us what port was selected
+            this._portRegistration = this._host.Services.GetRequiredService<IHostApplicationLifetime>().ApplicationStarted.Register(() =>
+            {
+                try
+                {
+                    var serverFeatures = this._host.Services.GetRequiredService<IServer>().Features;
+                    var addresses = serverFeatures.Get<IServerAddressesFeature>()!.Addresses.ToList();
+                    Debug.Assert(addresses.Count == 1, "Server should only be listening on one address");
+                    var uri = new Uri(addresses[0]);
```

Comment:
Wonder if this should just be:

```suggestion
                    // Server should only be listening on one address
                    var address = serverFeatures.Get<IServerAddressesFeature>()!.Addresses.Single();
                    var uri = new Uri(address);
```